File: src/Atlas/Battlenet/BinaryWriter.cs
using System.IO;
using System.Text;

namespace Atlas.Battlenet
{
    class BinaryWriter : System.IO.BinaryWriter
    {
        private Encoding _encoding = Encoding.ASCII;

        public BinaryWriter(Stream output) : base(output) { }
        public BinaryWriter(Stream output, Encoding encoding) : base(output, encoding) { }
        public BinaryWriter(Stream output, Encoding encoding, bool leaveOpen) : base(output, encoding, leaveOpen) { }

        public override void Write(string value)
        {
            if (value != null)
            {
                byte[] buffer = _encoding.GetBytes(value);
                Write(buffer);
            }
            Write((byte)0);
        }
    }
}

File: src/AtlasV/Battlenet/BNET/BinaryWriter.vb
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet
    Class BinaryWriter
        Inherits System.IO.BinaryWriter

        Public Sub New(ByVal output As Stream)
            MyBase.New(output, Encoding.UTF8)
        End Sub

        Public Sub New(ByVal output As Stream, ByVal encoding As Encoding)
            MyBase.New(output, encoding)
        End Sub

        Public Sub New(ByVal output As Stream, ByVal encoding As Encoding, ByVal leaveOpen As Boolean)
            MyBase.New(output, encoding, leaveOpen)
        End Sub

        Public Overrides Sub Write(ByVal value As String)
            If value IsNot Nothing Then Write(Encoding.UTF8.GetBytes(value))
            Write(CByte(0))
        End Sub

        Public Sub WriteByteString(ByVal value As Byte())
            If value IsNot Nothing Then Write(value)
            Write(CByte(0))
        End Sub
    End Class
End Namespace